### PR TITLE
kvserver: optimize check of required expiration leases

### PR DESCRIPTION
**Pull request by erikgrinaker**
Ranges located before or including the node liveness table must always use expiration leases to avoid circular dependencies on the node liveness table.

Previously, `requiresExpiringLeaseRLocked()` would determine whether this was the case for a given replica on-the-fly, by comparing the range descriptor start key with the node liveness table span. However, key comparisons are not cheap, which makes this method unsuitable for use in hot paths. This is already called for every RPC request via `checkRequestTimeRLocked()`, and further calls are planned with the introduction of `kv.expiration_leases_only.enabled`.

This patch instead makes this determination on range descriptor changes and stores the result in `Replica.mu.requiresExpirationLease`.

`requiresExpiringLeaseRLocked()` is also renamed to `requiresExpirationLeaseRLocked()`, which is the common name for such leases.

Touches #93903.

Epic: none
Release note: None

### CODE REVIEW

**Review by tbg — COMMENTED** [review]
Approach LGTM but I'll leave this one to Pavel since the salient question in this review is right up his alley - is there any other place that may change the key bounds of the descriptor that we're forgetting here?

**Comment by erikgrinaker** [comment]
> is there any other place that may change the key bounds of the descriptor that we're forgetting here?

The field description claims there shouldn't be, at least: https://github.com/cockroachdb/cockroach/blob/master/pkg/kv/kvserver/kvserverpb/state.pb.go#L51-L52

File: pkg/kv/kvserver/replica_init.go [hunk 1]
```diff
@@ -395,6 +395,15 @@ func (r *Replica) setDescLockedRaftMuLocked(ctx context.Context, desc *roachpb.R
 	r.concMgr.OnRangeDescUpdated(desc)
 	r.mu.state.Desc = desc
 
+	// Determine whether the replica unconditionally requires an expiration-based
+	// lease. Ranges located before or including the node liveness table must
+	// always use expiration leases to avoid circular dependencies on the node
+	// liveness table.
+	//
+	// Some tests don't use node liveness, and require expiration leases.
+	r.mu.requiresExpirationLease = r.store.cfg.NodeLiveness == nil ||
```
Comment (pavelkalinnikov) [inline pkg/kv/kvserver/replica_init.go:404]: I audited the code, `setDescLockedRaftMuLocked` is always invoked when we change the replica's `Desc`, so this seems indeed the only place where we should update `requiresExpirationLease`.

Technically, the assigning of `Desc` happens in other places, but it's alway coupled with `setDescLockedRaftMuLocked` in the same block.

**Review by pav-kv — APPROVED** [review]

**Comment by pavelkalinnikov** [comment]
Just curious: how come the key comparison (which, it looks like, boils down to `bytes.Compare`) is so expensive?

**Comment by erikgrinaker** [comment]
> Just curious: how come the key comparison (which, it looks like, boils down to `bytes.Compare`) is so expensive on a relative scale?

Byte comparisons, which boil down to a `memcmp`, are expensive enough that we avoid them where possible in storage code. Eliminating a byte comparison in e.g. a hot scan loop will often improve performance by something like 10%, see #85580 for an example.

The relative impact here on the request path is probably minor (certainly below 1%), but when I was working on a separate PR that checked for expiration leases in `shouldReplicaQuiesce` which is called on Raft ticks these comparisons showed up in profiles. They are also called for every RPC request, and will soon be called multiple times. I can run a quick benchmark just to see, but I wouldn't be surprised if it didn't register at all.

File: pkg/kv/kvserver/replica_init.go (hunk 1 above)
Comment (pavelkalinnikov) [reply pkg/kv/kvserver/replica_init.go:404]: I mean, there are a few places which assign the whole `r.mu.state` (which includes `Desc`), but only 2 cases are interesting: where `Desc` is initialized (which also means it has a key span). These are split and snapshot apply paths. In both of them loading `r.mu.state` is coupled with `setDescLockedRaftMuLocked `.

**Comment by erikgrinaker** [comment]
Added a quick `BenchmarkKV/Get` benchmark, doesn't register:

```
                        │ bench-before.txt │        bench-after.txt        │
                        │      sec/op      │   sec/op     vs base          │
KV/Get/Native/rows=1-24        34.45µ ± 3%   34.61µ ± 2%  ~ (p=0.579 n=10)
```

I'm fine with leaving this PR until we actually need the perf gains, e.g. if we need to start checking this in the Raft tick loop. if people feel strongly about it.

**Comment by pavelkalinnikov** [comment]
I don't feel strongly, up to you; correctness LGTM. Seems like complexity++ with low cost/risk, but also performance++ conditional to this being used heavily in the future (but maybe YAGNI, I'm trusting you to judge this now or later).

**Pull request closed without merging**